New Reconciler Extensions

mtesseract opened the pull request.

mtesseract committed:
Introduce extensions concept and implement pre-/post-hooks in terms of extensions.

mtesseract committed:
Introduce Extension points

mtesseract committed:
Inject logger into extensions

mtesseract committed:
Pass context to extensions

mtesseract committed:
Add PreDeletion extension point

mtesseract committed:
Pass Helm release to PreReconciliation Extension

mtesseract committed:
Refactoring

mtesseract committed:
Pass chartutil.Values to extensions

mtesseract committed:
Added docs

porridge commented on pkg/extension/extension.go:
```diff
@@ -0,0 +1,22 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type Extension interface{}
```
Since the Extension interface is completely unrelated to the specific extension types, I'm a little worried that compiler will not complain and an extension will simply be silently ignored in case someone has a typo in the method name.

Not sure if there is a good solution, but perhaps an example where you do something like `var _ PreReconciliationExtension = &myThing{}` idiom?

mtesseract replied:
I see the problem as well. I tend to believe that the advantages of this approach outweigh its disadvantages, though. Definitely a :+1: for the idea with the example.

porridge commented on pkg/extension/extension.go:
```diff
@@ -0,0 +1,22 @@
+type PreDeletionExtension interface {
+	PreDelete(ctx context.Context, obj *unstructured.Unstructured, release release.Release, vals chartutil.Values) error
+}
+
```
No need for a PostDelete extension? 🤔

mtesseract replied:
Not sure. I believe *we* don't have this use-case currently. Do you just have a feeling that it is missing or do you have a strong example? Besides, we can also ask Joe for his input on this.

But, the nice thing about this approach is that we can easily add it without breaking anything.

misberner replied:
It's easily conceivable that a `PostUninstall` is relevant. I'd rather add it upfront.

mtesseract replied:
ok

misberner commented on pkg/extension/extension.go:
```diff
@@ -0,0 +1,22 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type Extension interface{}
```
Hmm, an empty interface is useless from a static type checking perspective. Since there is no explicit implementation of an interface in Golang, the concept of a "marker interface" doesn't even make sense. There are some usages of `type Foo interface{}` in Go code, but these are rare, and I would say the drawbacks of losing static type checking outweigh the benefits.

If anything, I would propose to use this as syntactic sugar around strongly typed extension registration. That is, you have `WithPreReconciliationExtension`, `WithPreDeletionExtension` etc. which are all strongly typed. If `myFancyExtension` is an extension of both types, you would have to do
```
WithPreReconciliationExtension(myFancyExtension), WithPreDeletionExtension(myFancyExtension), ...
```

You can then add a level of syntactic sugar `WithExtension()`, that does the type checking at the time of _its_ invocation:
```
func WithExtension(e extension.Extension) Option {
	var registrations []Option
	if preReconcileExt, ok := e.(PreReconciliationExtension); ok {
		registrations = append(registrations, WithPreReconciliationExtension(preReconcileExt))
	}
	if preDeleteExt, ok := e.(PreDeletionExtension); ok { // no else if!
		registrations = append(registrations, WithPreDeletionExtension(preDeleteExt))
	}
	...
	if len(registrations) == 0 {
		panic(fmt.Sprintf("Extension of type %T does not implement any relevant interface", e))
	}
	return func(r *Reconciler) error {
		for _, reg := registrations {
			reg(r)
		}
	}
}
```
This makes sure we can check that somebody doesn't supply a bogus type at initialization time, instead of silently skipping over it at runtime. It also leaves a strongly typed interface for the common use case of having an extension of only a single type.

misberner replied:
I think I also don't get the motivation for the flexibility. What "new extension points" are you thinking of? The logic of the reconciler is really fairly simple, and the Helm-related actions should remain atomic from an extension point of view.

mtesseract replied:
I like the idea with using the strongly typed registration.

mtesseract replied:
Apropos "empty interface" -- I have actually been wondering if some `Name() string` method should be added to the `Extension` interface so that, for example, errors from the extension can be logged together with the failing extensions' name. WDYT?

mtesseract replied:
Regarding the "new extension points", this is actually something that Joe brought up. I have asked him about this, though.

misberner commented on pkg/extension/extension.go:
```diff
@@ -0,0 +1,22 @@
+
+type Extension interface{}
+
+type PreReconciliationExtension interface {
```
The nomenclature of pre/post reconciliation is flawed, as extension runs become _part_ of reconciliation (and deletion is _also_ part of reconciliation). What is really meant here is that they run before/after the respective helm steps (pre/post upgrade or install, pre uninstall - why no post uninstall, though?)

mtesseract replied:
Agreed, the naming should be changed.

mtesseract replied:
I do agree that the naming could be misleading, because of what you say: the extensions are supposed to be part of reconciliation -- with that I mean, part of the `Reconcile()` method of the reconciler. But, the term is somewhat overloaded in the internals of the helm-operator-plugins. As I mentioned elsewhere, the underlying `ActionInterface` has a certain set of actions, corresponding to the different flows within the reconciler and this set of actions also includes a `Reconcile()` method, along with the methods which directly correspond to Helm actions.

I tend to believe that what we are aiming for could be Post/Pre wrappings for each of these methods, i.e. Post/PreInstall, Post/PreUninstall,, Post/PreUpgrade, Post/PreReconcile.

I agree that the overloading here is not ideal, but ignoring this for a second, what do you think of this?

misberner commented on pkg/extension/extension.go:
```diff
@@ -0,0 +1,22 @@
+type PreReconciliationExtension interface {
+	PreReconcile(ctx context.Context, obj *unstructured.Unstructured, release release.Release, vals chartutil.Values) error
+}
+type PreDeletionExtension interface {
```
There also need to be comments to explain the deletion workflow. In particular, it should be explained that if a CR exists but has a non-null deletion timestamp, `PreReconcile` above will _not_ be invoked, but `PreDelete` will (I'm btw still not convinced that the distinction between `PreReconcile` and `PreDelete` as potentially independent things makes sense - a correctly implemented extension not leaking resources that implements `PreReconcile` but doesn't need `PreDelete` seems not impossible, but very unlikely -, but I don't know what conversations you had with Joe Lanford to that end)

misberner commented on pkg/reconciler/extensions.go:
```diff
@@ -0,0 +1,75 @@
+	es.extensions = append(es.extensions, e)
+}
+
+func (es *extensions) len() int {
+	return len(es.extensions)
+}
+
+func (es *extensions) get(idx int) extension.Extension {
```
Why do these exist? They are unexported, so every caller that can access these methods can also access the `extensions` field directly

mtesseract replied:
My idea here was to encapsulate the extensions registration/execution.

misberner replied:
I mean `get` and `len`. I don't think an unexported method should be regarded as in any way privileged over an unexported field when it comes to encapsulation/interfaces.

misberner commented on pkg/reconciler/extensions.go:
```diff
@@ -0,0 +1,75 @@
+	return es.extensions[idx]
+}
+
+func (es *extensions) iterate(f func(e extension.Extension) error) error {
```
Same question here

mtesseract replied:
To reduce code duplication for calling each extension.

misberner replied:
Makes sense, but I would call it `forEach` or so.

mtesseract replied:
ok

misberner commented on pkg/reconciler/extensions.go:
```diff
@@ -0,0 +1,75 @@
+	return err
+}
+
+func (es *extensions) loggerInto(l logr.Logger) {
```
The method name doesn't make sense. The function below is to be read as "inject logger into", by taking the package name into account. And "into" at the end of a function is commonly read as to refer to the (last) argument of the function call. So if we need it, it should be called `injectLoggerIntoAll`

mtesseract replied:
👍

misberner commented on pkg/reconciler/reconciler.go:
```diff
@@ -525,7 +529,22 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
+		release = *rel
+	}
+
+	if obj.GetDeletionTimestamp() != nil {
```
This can't be moved here. If you are in a state where `r.getValues` returns an error (and this might happen simply from a misconfiguration at the CR level, such as referencing a secret that doesn't exist, resulting in us not being able to build the values due to missing secret content), you will find the controller to be unable to perform a deletion. Deletions must strictly be handled _before_ any non-internal error (such as not being able to get the action client) might terminate the control flow.

mtesseract replied:
Get it. I guess this would then imply that the PreDeletion extension should neither receive `vals`, nor `release`, right?

[On the other hand, one could also introduce a fallback mechanism where `vals` is simply passed as empty to the `PreDelete` extension in case an error occurs while assembling it -- similar to the situation with `release`].

misberner commented on pkg/reconciler/reconciler.go:
```diff
@@ -537,10 +556,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
-		if err := h.Exec(obj, vals, log); err != nil {
-			log.Error(err, "pre-release hook failed")
-		}
+	r.extensions.loggerInto(log)
```
I don't think I understand why injecting loggers is needed, especially at runtime on each reconciliation. If it's about compatibility with the old hook interface, any retrofitting of interfaces should IMHO happen exactly once, namely at registration time. Meaning that when you call `WithPreHook`, the hook function is converted into an extension and that extension is then registered. Subsequent logic will treat this extension as every other extension.

mtesseract replied:
Well, Joe suggested to use an injection API as it is currently being done within controller-runtime. And the thing with the `Reconcile()` method is that it currently derives a new logger from the existing one with a special `WithValues()` configuration based on the `req` parameter of the `Reconcile()` function. :-/

misberner commented on pkg/reconciler/reconciler.go:
```diff
@@ -537,10 +556,11 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.
+
+	err = r.extPreReconcile(ctx, obj, release, vals)
+	if err != nil {
+		log.Error(err, "pre-release hook failed")
```
No return? This means that if an extension fails, we'll not try again after a back-off. This limitation of pre hooks was part of the reason why I introduced extensions in the first place. Again, if hook functions require special retrofitting (ignoring returned errors), this should happen at registration time - meaning your hook wrapper extension is responsible of logging the error, if any, and then unconditionally returning a `nil` error from `PreReconcile`.

mtesseract replied:
The return is missing, thanks.

misberner replied:
It's not just a missing return. What's the point of logging if you're returning the error anyway? Also, if the aim of the hook wrappers is to be backwards-compatible, then you need to move the logging and the `return nil` to the respective methods. Otherwise you would change the behavior.

mtesseract replied:
changed, pushed

mtesseract committed:
WIP

Add test

mtesseract committed:
Remove release, values arguments from PreReconcile extension

mtesseract committed:
Pass logger via context

mtesseract committed:
Refactor extension interface

mtesseract committed:
Simplify extensions.

mtesseract committed:
Introduce a 'Reconciliation Context'

mtesseract committed:
Added nil safe getter for kubernetes config

mtesseract committed:
Simplfy hooks

mtesseract committed:
Rename extensions.
Make sure to not forget running EndReconciliation extension in the deletion case.

mtesseract committed:
Polishing

mtesseract committed:
Support named extensions
